x-pack/plugins/case/server/common/utils.ts

```diff
@@ -0,0 +1,125 @@
+ * Default unknown user
+ */
+export const nullUser: User = { username: null, full_name: null, email: null };
+
+/**
+ * Adds the ids and indices to a map of statuses
+ */
+export function addAlertInfoToStatusMap({
+  comment,
+  statusMap,
+  status,
+}: {
+  comment: CommentAttributes;
+  statusMap: Map<CaseStatuses, AlertInfo>;
+  status: CaseStatuses;
+}) {
+  const newAlertInfo = getAlertIndicesAndIDs([comment]);
+
+  // combine the already accumulated ids and indices with the new ones from this alert comment
+  if (newAlertInfo.ids.length > 0 && newAlertInfo.indices.size > 0) {
+    const accAlertInfo = statusMap.get(status) ?? { ids: [], indices: new Set<string>() };
+    accAlertInfo.ids.push(...newAlertInfo.ids);
+    accAlertInfo.indices = new Set<string>([
+      ...accAlertInfo.indices.values(),
+      ...newAlertInfo.indices.values(),
+    ]);
+    statusMap.set(status, accAlertInfo);
+  }
+}
+
+/**
+ * Combines multiple filter expressions using the specified operator and parenthesis if multiple expressions exist.
+ * This will ignore empty string filters. If a single valid filter is found it will not wrap in parenthesis.
+ *
+ * @param filters an array of filters to combine using the specified operator
+ * @param operator AND or OR
+ */
+export const combineFilters = (filters: string[] | undefined, operator: 'OR' | 'AND'): string => {
+  const noEmptyStrings = filters?.filter((value) => value !== '');
+  const joinedExp = noEmptyStrings?.join(` ${operator} `);
+  // if undefined or an empty string
+  if (!joinedExp) {
+    return '';
+  } else if ((noEmptyStrings?.length ?? 0) > 1) {
+    // if there were multiple filters, wrap them in ()
+    return `(${joinedExp})`;
+  } else {
+    // return a single value not wrapped in ()
+    return joinedExp;
+  }
+};
+
+/**
+ * Counts the total alert IDs within a single comment.
+ */
+export const countAlerts = (comment: SavedObjectsFindResult<CommentAttributes>) => {
+  let totalAlerts = 0;
+  if (
+    comment.attributes.type === CommentType.alert ||
+    comment.attributes.type === CommentType.generatedAlert
+  ) {
+    if (Array.isArray(comment.attributes.alertId)) {
+      totalAlerts += comment.attributes.alertId.length;
+    } else {
+      totalAlerts++;
+    }
+  }
+  return totalAlerts;
+};
+
+/**
+ * Count the number of alerts for each id in the alert's references. This will result
+ * in a map with entries for both the collection and the individual sub cases. So the resulting
+ * size of the map will not equal the total number of sub cases.
+ */
+export const groupTotalAlertsByID = ({
+  comments,
+}: {
+  comments: SavedObjectsFindResponse<CommentAttributes>;
+}): Map<string, number> => {
+  return comments.saved_objects.reduce((acc, alertsInfo) => {
+    for (const alert of alertsInfo.references) {
+      if (alert.id) {
+        const totalAlerts = acc.get(alert.id);
+        if (totalAlerts !== undefined) {
+          acc.set(alert.id, totalAlerts + countAlerts(alertsInfo));
```

Comment:
You can cache countAlerts `const alertCount = countAlerts(alertsInfo)` and use that on 105 maybe?

Reply:
Good point, will update